server.go

```diff
@@ -160,6 +171,500 @@ func (s *server) RemoveRebroadcastInventory(iv *wire.InvVect) {
+	// being disconnected) and wasting memory.
+	s.blockManager.QueueTx(tx, p)
+	s.blockManager.WaitForTx()
+}
+
+// handleBlockMsg is invoked when a peer receives a block bitcoin message.
+// It blocks until the bitcoin block has been fully processed.
+func (s *server) handleBlockMsg(p *peer.Peer, msg *wire.MsgBlock, buf []byte) {
+	// Convert the raw MsgBlock to a btcutil.Block which provides some
+	// convenience methods and things such as hash caching.
+	block := btcutil.NewBlockFromBlockAndBytes(msg, buf)
+
+	// Add the block to the known inventory for the peer.
+	iv := wire.NewInvVect(wire.InvTypeBlock, block.Sha())
+	p.AddKnownInventory(iv)
+
+	// Queue the block up to be handled by the block
+	// manager and intentionally block further receives
+	// until the bitcoin block is fully processed and known
+	// good or bad.  This helps prevent a malicious peer
+	// from queueing up a bunch of bad blocks before
+	// disconnecting (or being disconnected) and wasting
+	// memory.  Additionally, this behavior is depended on
+	// by at least the block acceptance test tool as the
+	// reference implementation processes blocks in the same
+	// thread and therefore blocks further messages until
+	// the bitcoin block has been fully processed.
+	s.blockManager.QueueBlock(block, p)
+	s.blockManager.WaitForBlock()
+}
+
+// handleInvMsg is invoked when a peer receives an inv bitcoin message and
+// is used to examine the inventory being advertised by the remote peer and
+// react accordingly.  We pass the message down to blockmanager which will
+// call QueueMessage with any appropriate responses.
+func (s *server) handleInvMsg(p *peer.Peer, msg *wire.MsgInv) {
+	s.blockManager.QueueInv(msg, p)
+}
+
+// handleHeadersMsg is invoked when a peer receives a headers bitcoin
+// message.  The message is passed down to the block manager.
+func (s *server) handleHeadersMsg(p *peer.Peer, msg *wire.MsgHeaders) {
+	s.blockManager.QueueHeaders(msg, p)
+}
+
+// handleGetData is invoked when a peer receives a getdata bitcoin message
+// and is used to deliver block and transaction information.
+func (s *server) handleGetDataMsg(p *peer.Peer, msg *wire.MsgGetData) {
+	numAdded := 0
+	notFound := wire.NewMsgNotFound()
+
+	// We wait on this wait channel periodically to prevent queueing
+	// far more data than we can send in a reasonable time, wasting memory.
+	// The waiting occurs after the database fetch for the next one to
+	// provide a little pipelining.
+	var waitChan chan struct{}
+	doneChan := make(chan struct{}, 1)
+
+	for i, iv := range msg.InvList {
+		var c chan struct{}
+		// If this will be the last message we send.
+		if i == len(msg.InvList)-1 && len(notFound.InvList) == 0 {
+			c = doneChan
+		} else if (i+1)%3 == 0 {
+			// Buffered so as to not make the send goroutine block.
+			c = make(chan struct{}, 1)
+		}
+		var err error
+		switch iv.Type {
+		case wire.InvTypeTx:
+			err = pushTxMsg(p, s, &iv.Hash, c, waitChan)
```

Comment:
With the aforementioned changes, this would become:

```Go
s.pushTxMsg(p, &iv.Hash, c, waitChan)
```

The same for the `pushBlockMsg` and `pushMerkleBlockMsg` below.